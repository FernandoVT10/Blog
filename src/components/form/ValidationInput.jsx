import { useState } from "react";

export default ({ type, placeholder, onChange, maxLength="100", prefix="" }) => {
    const [error, setError] = useState("");

    const validate = ({ target: { value } }) => {
        if(type === "email") {
            if(!/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(value)) {
                onChange({value, valid: false});
                setError("The email is invalid");
                return;
            }
        }

        onChange({value, valid: true});
        setError("");
    };

    const inputClass = error ? "formulary__input--error" : "";
    const labelClass = error ? "formulary__error-label--error" : "";

    return (
        <div className="formulary__input-container">
            <input
            id={`${prefix}-input`}
            type={type}
            className={`formulary__input ${inputClass}`}
            onChange={validate}
            placeholder={placeholder}
            autoComplete={type}
            maxLength={maxLength}
            required/>

            <label
            className={`formulary__error-label mt-2 ${labelClass}`}
            htmlFor={`${prefix}-input`}>
                <i className="fas fa-info-circle mr-1" aria-hidden="true"></i>

                { error }
            </label>
        </div>
    );
};