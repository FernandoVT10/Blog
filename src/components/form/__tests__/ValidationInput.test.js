import ValidationInput from "../ValidationInput";
import { act, Simulate } from "react-dom/test-utils";
import { render } from "react-dom";

let container;

describe("<ValidationInput/> component", () => {
    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
    });

    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    it("Check if renders correctly", async () => {
        await act(async () => {
            render(
                <ValidationInput
                type="email"
                placeholder="Test input"
                onChange={() => {}}/>,
            container);
        });
    
        expect(container).toMatchSnapshot();
    });

    it("It should validate the email", async () => {
        const onChange = jest.fn();

        await act(async () => {
            render(
                <ValidationInput
                type="email"
                placeholder="Test input"
                onChange={onChange}/>,
            container);
        });

        const input = container.querySelector("input");
        const label = container.querySelector("label");

        act(() => {
            input.value = "test@ok.a";
            Simulate.change(input);
        });

        expect(label.textContent).toBe("The email is invalid");
        expect(onChange).toHaveBeenCalledWith({ value: "test@ok.a", valid: false });

        act(() => {
            input.value = "test@gmail.com";
            Simulate.change(input);
        });

        expect(label.textContent).toBe("");
        expect(onChange).toHaveBeenCalledWith({ value: "test@gmail.com", valid: true });
    });
});