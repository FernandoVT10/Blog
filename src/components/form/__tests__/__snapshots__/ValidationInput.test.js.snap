// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ValidationInput/> component Check if renders correctly 1`] = `
<div>
  <div
    class="formulary__input-container"
  >
    <input
      autocomplete="email"
      class="formulary__input "
      id="-input"
      maxlength="100"
      placeholder="Test input"
      required=""
      type="email"
    />
    <label
      class="formulary__error-label mt-2 "
      for="-input"
    >
      <i
        aria-hidden="true"
        class="fas fa-info-circle mr-1"
      />
      
    </label>
  </div>
</div>
`;
